addons/stock/report/product_label_report.py

```diff
@@ -22,11 +29,40 @@ def _get_report_values(self, docids, data):
         quantity_by_product = defaultdict(list)
         for p, q in data.get('quantity_by_product').items():
             product = Product.browse(int(p))
-            quantity_by_product[product].append((product.barcode, q))
+            product_info = {
+                'barcode': product.barcode,
+                'quantity': q,
+                'display_name_markup': markupsafe.Markup(product.display_name),
+                'default_code': self._get_markup_default_code(product)
+            }
+            quantity_by_product[product].append(product_info)
         if data.get('custom_barcodes'):
             # we expect custom barcodes to be: {product: [(barcode, qty_of_barcode)]}
             for product, barcodes_qtys in data.get('custom_barcodes').items():
-                quantity_by_product[Product.browse(int(product))] += (barcodes_qtys)
+                product = Product.browse(int(product))
+                barcodes_qtys[0] = {
+                    'barcode': barcodes_qtys[0][0],
+                    'quantity': barcodes_qtys[0][1],
+                    'display_name_markup': markupsafe.Markup(product.display_name),
+                    'default_code': self._get_markup_default_code(product)
+                }
```

Comment:
As hinted by the plural form of `barcodes_qtys`, it is a list. Meaning that there could be more than a single lot in there. :wink: